upstream: add limits to connection pool map

klarose opened the pull request:
*Description*: Add limits to the connection pool map in preparation for actually configuring them. They default to disabled, and there is no way yet to configure them in a production build, so there should be no impact to the actual system right now. However, I have done extensive UT of the new code, so it should work fairly well.

The final steps in this process to get it actually working will be to implement a method to configure the limits, and handle them failing in `ClusterManagerImpl`.

*Risk Level*: Medium. Should be no functional change, but it's in the main http path.
*Testing*: Lots of UT
*Docs Changes*: None
*Release Notes*: None

klarose commented:
/retest

klarose commented:
@mattklein123 This is a continuation of the previous PR which added the ConnPoolMap. PTAL if you can. Thanks!

klarose commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -1120,13 +1121,13 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
   });
-
-  return &pool;
+  RELEASE_ASSERT(pool.has_value(), "Pool allocation should never fail");
```
I'm still not 100% clear on when `RELEASE_ASSERT` makes sense... I did it here because I was concerned there could be corner cases in the near future that aren't caught in testing, but would be better found explicitly in production rather than leading to undefined behaviour. I can change it to a normal `ASSERT` if that's wrong.

mattklein123 replied:
Yes I would use `ASSERT` here as this is a programming error. But if it can't fail, why is this returning an optional at all? I guess it's because it can't happen right now given you can't configure limits? Yeah I would just ASSERT and add a comment.

klarose replied:
Yeah. I wanted to make it clear that it's not expected to fail right now in case somebody was working in this area while I work on the next phase. The assert will hopefully be gone in the next week or two, but you never know.

mattklein123 commented on source/common/upstream/conn_pool_map.h:
```diff
@@ -51,10 +54,21 @@ template <typename KEY_TYPE, typename POOL_TYPE> class ConnPoolMap {
   Event::Dispatcher& thread_local_dispatcher_;
   std::vector<DrainedCb> cached_callbacks_;
   Common::DebugRecursionChecker recursion_checker_;
+  absl::optional<uint64_t> max_size_;
```
nit: const

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -6,30 +6,42 @@ namespace Envoy {
+
+  // We need a new element. If we can, create a pool and add it to the map. Tell it about any cached
+  // callbacks.
+  if (max_size_.has_value() && size() >= max_size_) {
```
nit: `max_size_. value()` (I can never remember if without it means the value or just whether it has a value)

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -6,30 +6,42 @@ namespace Envoy {
+  if (max_size_.has_value() && size() >= max_size_) {
+    registerFullCallbacks();
+    if (!cleanDrainedPools()) {
+      // nothing free. :( Return failure.
```
nit: "Nothing free. Return failure."

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -63,5 +75,38 @@ void ConnPoolMap<KEY_TYPE, POOL_TYPE>::drainConnections() {
+}
+
+template <typename KEY_TYPE, typename POOL_TYPE>
+void ConnPoolMap<KEY_TYPE, POOL_TYPE>::registerFullCallbacks() {
```
This is only called from one place, right? Can you inline it there and add more comments on what the logic is doing? It's not immediately obvious from a quick read.

klarose replied:
I typically do this (take loops, conditions, etc) out into their own functions as a method of reducing cyclomatic complexity and depth-of-nesting -- I find it leads to more maintainable and readable code. If you'd prefer it inlined I can certainly do that. In fact, it was inlined originally. I only moved it out on a second pass. :)

Either way, it can definitely use more comments.

mattklein123 replied:
Either way is fine, just please add some more comments and I can take another pass.

klarose commented:
@mattklein123 Thanks for the review. I've addressed your comments. The logic could be simplified if we changed how the callbacks worked, or enhanced them slightly. For example, if we could make the immediate drained callback optional, then expose a function to indicate whether the pool is drained, we could do away with the guards and the temporary list of drained pools.

I'm not sure we want to do that level of refactoring here, though.

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -6,30 +6,54 @@ namespace Envoy {
+  if (max_size_.has_value() && size() >= max_size_.value()) {
+    // We have hit the maximum number of connection pools. Ask the tracked pools to inform us when
+    // we can free them.
+    registerFullCallbacks();
```
I think I better understand what is going on here now. Thanks for the comments. Q: Can all of this be simplified by doing the following: Loop over all pools, and instead of registering callbacks, just have some method called something like `hasActiveConnections()` or something like that, and if not, just delete it inline? I think if we did this, a lot of things become a lot simpler, we just get rid of the first one we find, no need for another vector, etc. and also then we don't leave callbacks installed after we do this operation which seems not optimal and potentially will lead to strange edge case problems?

klarose replied:
That sounds like a good approach. We'll have the same semantics -- we know right away if there's one free, and if there isn't we'll fail. If one frees up later, then the next request will work.

The only downside I can see to this is that if there are a large number of connection pools, and we're at the limit with pools only freeing up occasionally, we'll end up O(n) iterating over them for each new request. This could be slow for a large limit. Of course, the current code I have up doesn't really solve this for that case either. My original design here had us tracking the "idle" state of a pool so we could have a ready-to-go list of pools which could be freed.

For this first round, we definitely don't need that. I'd  rather build the rest of the solution out and then optimize it.

klarose commented:
@mattklein123 Thanks again. It's quite a bit simpler now. I'm glad you suggested that change.

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -6,30 +6,50 @@ namespace Envoy {
+
+  // We need a new pool. Check if we have room.
+  if (max_size_.has_value() && size() >= max_size_.value()) {
+
```
nit: del newline

mattklein123 commented on source/common/upstream/conn_pool_map_impl.h:
```diff
@@ -6,30 +6,50 @@ namespace Envoy {
+
+    // We're full. Try to free up a pool. If we can't, bail out.
+    if (!freeOnePool()) {
+      // Nothing free. Return failure.
```
nit: this comment is redundant with the one just above, I would delete.

mattklein123 commented on test/common/http/http1/conn_pool_test.cc:
```diff
@@ -721,6 +721,50 @@ TEST_F(Http1ConnPoolImplTest, RemoteCloseToCompleteResponse) {
+}
+
+TEST_F(Http1ConnPoolImplTest, ActiveRequestHasActiveConnectionsTrue) {
+
```
nit: del newline. Same elsewhere.

mattklein123 commented on test/common/http/http2/conn_pool_test.cc:
```diff
@@ -114,6 +114,20 @@ class Http2ConnPoolImplTest : public TestBase {
   // Asserts that the provided requests receives onPoolFailure.
   void expectStreamReset(ActiveTestRequest& r);
 
+  /** Closes a test client
```
nit: formatting, also please newline between method and the next comment, same below.
